toolchain/check/function.cpp

```diff
@@ -0,0 +1,216 @@
+}
+
+// Returns true if the provided function signatures agrees, in the sense that
+// declarations can be merged.
+static auto DoesFunctionSignatureAgree(Context& context,
+                                       const SemIR::Function& new_function,
+                                       const SemIR::Function& prev_function)
+    -> bool {
+  if (FunctionSignatureHasError(context, new_function) ||
+      FunctionSignatureHasError(context, prev_function)) {
+    return false;
+  }
+  if (!DoParamAgrees(context, new_function.decl_id,
+                     new_function.implicit_param_refs_id, prev_function.decl_id,
+                     prev_function.implicit_param_refs_id, " implicit") ||
+      !DoParamAgrees(context, new_function.decl_id, new_function.param_refs_id,
+                     prev_function.decl_id, prev_function.param_refs_id, "")) {
+    return false;
+  }
+  if (new_function.return_type_id != prev_function.return_type_id) {
+    CARBON_DIAGNOSTIC(
+        FunctionSignatureReturnTypeDisagree, Error,
+        "Function declaration with return type of `{0}` disagrees.",
+        SemIR::TypeId);
+    CARBON_DIAGNOSTIC(FunctionSignatureReturnTypeDisagreeNoReturn, Error,
+                      "Function declaration with no return type disagrees.");
+    auto diag =
+        new_function.return_type_id.is_valid()
+            ? context.emitter().Build(new_function.decl_id,
+                                      FunctionSignatureReturnTypeDisagree,
+                                      new_function.return_type_id)
+            : context.emitter().Build(
+                  new_function.decl_id,
+                  FunctionSignatureReturnTypeDisagreeNoReturn);
+    if (prev_function.return_type_id.is_valid()) {
+      CARBON_DIAGNOSTIC(FunctionSignatureReturnTypePrevious, Note,
+                        "Matched function declaration with return type `{0}`.",
+                        SemIR::TypeId);
+      diag.Note(prev_function.decl_id, FunctionSignatureReturnTypePrevious,
+                prev_function.return_type_id);
+    } else {
+      CARBON_DIAGNOSTIC(FunctionSignatureReturnTypePreviousNoReturn, Note,
+                        "Matched function declaration with no return type.");
+      diag.Note(prev_function.decl_id,
+                FunctionSignatureReturnTypePreviousNoReturn);
+    }
+    diag.Emit();
+    return false;
+  }
+
+  return true;
+}
+
+auto MergeFunctionDecl(Context& context, Parse::NodeId parse_node,
+                       SemIR::Function& new_function,
+                       SemIR::FunctionId prev_function_id, bool is_definition)
+    -> bool {
+  auto& prev_function = context.functions().Get(prev_function_id);
+
+  // TODO: Disallow redeclarations within classes?
+  if (!DoesFunctionSignatureAgree(context, new_function, prev_function)) {
+    return false;
+  }
+
+  if (!is_definition) {
+    CARBON_DIAGNOSTIC(FunctionRedeclaration, Error,
+                      "Redeclaration of function {0}.", SemIR::NameId);
```

Comment:
```suggestion
                      "Redundant redeclaration of function {0}.", SemIR::NameId);
```
I think this'll make it a little clearer why this is an error.

Reply:
Done